File: qna/src/main/java/com/example/rag/qna/domain/cache/QuestionCache.java
package com.example.rag.qna.domain.cache;

import core.rag.Question;
import org.springframework.stereotype.Component;

import java.util.LinkedHashMap;
import java.util.Map;

@Component
public class QuestionCache {
    private final Map<String, Question> map;
    private final int maxSize;

    public QuestionCache() {
        this.maxSize = 1000;
        this.map = new LinkedHashMap<>(1000, 0.75f, true);
    }

    public Question getRecentQuestion(String questionId) {
        return map.get(questionId);
    }

    //  FIFO 구조
    public void putQuestion(String questionId, Question question) {
        // 비효율 - 개선 사항
        if (map.size() >= maxSize) {
            String oldestKey = map.keySet().iterator().next();
            map.remove(oldestKey);
        }
        map.put(questionId, question);
    }
}

File: qna/src/main/kotlin/com/example/rag/qna/domain/cache/QuestionCache.kt
package com.example.rag.qna.domain.cache

import core.rag.Question
import org.springframework.stereotype.Component

@Component
class QuestionCache {
    private val maxSize = 1000
    private val map: MutableMap<String, Question> = LinkedHashMap(maxSize, 0.75f, true)

    fun getRecentQuestion(questionId: String): Question? {
        return map[questionId]
    }

    //  FIFO 구조
    fun putQuestion(questionId: String, question: Question) {
        // 비효율 - 개선 사항
        if (map.size >= maxSize) {
            val oldestKey = map.keys.iterator().next()
            map.remove(oldestKey)
        }
        map[questionId] = question
    }
}